### PR TITLE
Add YOLO PreProcessor

### PR DESCRIPTION
**Pull request by kartikdutt18**
Hey everyone,
This PR aims to add preprocessor function for YOLO class. This basically converts annotations into feature maps that can be used for training YOLO model i.e. have the same format as output of YOLO model.
To Do :
- [x] Add tests to verify output.

CC : @KimSangYeon-DGU, @saksham189, @zoq

### CODE REVIEW

File: dataloader/preprocessor.hpp [hunk 1]
```diff
@@ -48,12 +48,112 @@ class PreProcessor
...
+          annotations(0, boxIdx).n_rows - 1), arma::span());
+      arma::vec labels = annotations(0, boxIdx).row(0);
+
+      // Normalize the coordinates.
+      boundingBoxes.row(0) /= imageWidth;
+      boundingBoxes.row(2) /= imageWidth;
+      boundingBoxes.row(1) /= imageHeight;
+      boundingBoxes.row(3) /= imageHeight;
+
+      // Get width and height as well as centres for the bounding box.
```
Comment (zoq) [inline dataloader/preprocessor.hpp:118]: Should this be "[...] as well as the center for the bounding box."?

Comment (kartikdutt18) [reply dataloader/preprocessor.hpp:118]: Right, I will fix it in the next commit.

File: dataloader/preprocessor.hpp [hunk 2]
```diff
@@ -48,12 +48,111 @@ class PreProcessor
...
+      // Get the bounding box and labels corresponding to current image.
+      arma::mat boundingBoxes = annotations(0, boxIdx)(arma::span(1,
+          annotations(0, boxIdx).n_rows - 1), arma::span());
+      arma::vec labels = annotations(0, boxIdx).row(0);
+
+      // Normalize the coordinates.
+      boundingBoxes.row(0) /= imageWidth;
+      boundingBoxes.row(2) /= imageWidth;
+      boundingBoxes.row(1) /= imageHeight;
+      boundingBoxes.row(3) /= imageHeight;
```
Comment (KimSangYeon-DGU) [inline dataloader/preprocessor.hpp:115]: Hey @kartikdutt18 
To normalize the width and height of image into -1 and 1, can you refer to [this](https://stats.stackexchange.com/questions/178626/how-to-normalize-data-between-1-and-1)?

Comment (KimSangYeon-DGU) [reply dataloader/preprocessor.hpp:115]: Ahh, I misunderstood the question

**Comment by saksham189** [comment]
Hey @kartikdutt18 , I just wanted to know whether you were stuck anywhere with this because we were planning to finish this first.

**Comment by kartikdutt18** [comment]
Hey @saksham189, I've probably written some expression wrong somewhere and I wasn't able to track it down before Saturday so as I implemented the weight translator [here](https://github.com/kartikdutt18/mlpack-PyTorch-Weight-Translator) so we can now transfer weights of Darknet model or any other model from PyTorch to mlpack. However other than weights we also need to transfer running mean and variance for BatchNorm Layer and with that I'm stuck with the error [here](https://github.com/mlpack/mlpack/pull/2519#issuecomment-661120039). I'll take a look at the preprocessor again today. In the meanwhile, I tried training Darknet 19 model on Cifar10, It didn't converge, so I also added Darknet 1 (same as YOLO model) on Cifar10 to see if we can train that and it takes 3 hours for a single epoch so we should also be able to train YOLO model (in less time than Darknet), if needed. Here is the log for the first epoch (20% accuracy).
21_log_epoch

**Comment by saksham189** [comment]
Looks good and the training time is also considerably lower. So, maybe we should just train that first and look at the accuracy.

**Comment by kartikdutt18** [comment]
> Looks good and the training time is also considerably lower. So, maybe we should just train that first and look at the accuracy.

Great, I'm letting the model train. In the meanwhile, I will try to get running mean and variance also loaded into mlpack. Also, with transferred weights we can also train the model on imagenette (that way running mean and variance will be set in training).

**Comment by kartikdutt18** [comment]
Hey @saksham189, I fixed the errors in the implementation as well added tests fro single image, multiple images and image with multiple bounding box.

File: tests/preprocessor_tests.cpp [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  bBox << 7 << 157 << 90 << 486 << 372 << arma::endr;
+  input(0, 2) = bBox;
+
+  PreProcessor<arma::mat, arma::field<arma::vec>>::YOLOPreProcessor(
+      input, output, 500, 387);
+
+  arma::vec desiredSum(3);
+  desiredSum << 8.3342 << 18.4093 << 7.13195 << arma::endr;
+  for (size_t i = 0; i < output.n_cols; i++)
+    BOOST_REQUIRE_CLOSE(arma::accu(output.col(i)), desiredSum(i), 1e-3);
```
Comment (zoq) [inline tests/preprocessor_tests.cpp:58]: Maybe we should also compare the normalized boxes with pre calculated ones, what do you think?

File: dataloader/preprocessor.hpp [hunk 4]
```diff
@@ -48,12 +48,117 @@ class PreProcessor
...
+      // Get the bounding box and labels corresponding to current image.
+      arma::mat labels(1, annotations(0, boxIdx).n_elem / 5);
+      arma::mat boundingBoxes(4, annotations(0, boxIdx).n_elem / 5);
+      for (size_t i = 0; i < boundingBoxes.n_cols; i++)
+      {
+        labels.col(i)(0) = annotations(0, boxIdx)(i * 5);
+        boundingBoxes.col(i) = annotations(0, boxIdx)(arma::span(i * 5 + 1,
+            (i + 1) * 5 - 1));
+      }
+
```
Comment (zoq) [inline dataloader/preprocessor.hpp:115]:
```suggestion
```

Looks like we can remove the extra line here.

File: dataloader/preprocessor.hpp [hunk 5]
```diff
@@ -48,12 +48,117 @@ class PreProcessor
...
+
+      // Normalize the coordinates.
+      boundingBoxes.row(0) /= imageWidth;
+      boundingBoxes.row(2) /= imageWidth;
+      boundingBoxes.row(1) /= imageHeight;
+      boundingBoxes.row(3) /= imageHeight;
+
+      // Get width and height as well as centres for the bounding box.
+      arma::mat widthAndHeight(2, boundingBoxes.n_cols);
```
Comment (zoq) [inline dataloader/preprocessor.hpp:124]: I have seen at least one dataset that alreay used the centered bbox format but wasn't normalized, do you think we should introduce a parameter to skip the step?

File: tests/preprocessor_tests.cpp (hunk 3 above)
Comment (kartikdutt18) [reply tests/preprocessor_tests.cpp:58]: Sure, I can add that as well.

File: dataloader/preprocessor.hpp (hunk 4 above)
Comment (kartikdutt18) [reply dataloader/preprocessor.hpp:115]: Makes sense. Thanks for the suggestion.

File: dataloader/preprocessor.hpp (hunk 5 above)
Comment (kartikdutt18) [reply dataloader/preprocessor.hpp:124]: Ahh makes sense. I'll add that as well.

**Comment by kartikdutt18** [comment]
>  So, maybe we should just train that first and look at the accuracy.

Hey @saksham189, @KimSangYeon-DGU, I continued the training. Initially it was for 5 epochs (the learning rate was 0.01) and after 5 epochs it remained constant and the below graph was obtained. Then,  I set learning rate to 0.001 and it's still training and has completed nearly 25 epochs out of 40. The training and validation accuracy are continuously increasing from 20% and 7% respectively to 31% and roughly 20%. The graph for that is also attached below. Also, about using larger image sizes such as imagenette, The training time doesn't change significantly with larger images however time for loading the dataset is very large. Imagenette with 320 x 320 x 3 image size and 5k images nearly took an hour.
**Current training log (lr = 1e-3).**
darknet_v1_fast_mean_log_epoch
darknet_v1_fast
 
**First attempt to train (5 epochs / lr = 1e-2) (training accuracy = 26%, Graphs are shown below)**
darknet_v1_slow_mean_log_epoch
darknet_v1_slow
I'll add the changes for YOLOv3 in a couple of minutes. Then I would also have to add tests for the same.

File: dataloader/preprocessor.hpp [hunk 6]
```diff
@@ -93,6 +93,174 @@ class PreProcessor
...
+                               const size_t imageHeight = 224,
+                               const size_t gridWidth = 7,
+                               const size_t gridHeight = 7,
+                               const size_t numBoxes = 2,
+                               const size_t numClasses = 20,
+                               const bool normalize = true)
+  {
+    // See if we can change this to v4 / v5.
+    mlpack::Log::Assert(version >= 1 && version <= 3, "Supported YOLO versions \
+        are version 1 to version 3.");
```
[the remark this answers is not in the document]
Comment (kartikdutt18) [reply dataloader/preprocessor.hpp:130]: Yes it does. It might also work for YOLOv4 unless they changed feature map but it works for all versions between 1 and 3 (inclusive).

**Comment by kartikdutt18** [comment]
I think this is ready to go as well.

File: dataloader/preprocessor.hpp [hunk 7]
```diff
@@ -93,6 +93,173 @@ class PreProcessor
...
   }
+
+  /**
+   * PreProcessor for YOLO model. Converts arma::field type annotations to
+   * arma::mat type for training YOLO model. Each column in target matrix has
+   * the size : gridWidth * gridHeight * (5 * numBoxes + classes).
+   *
+   * @param annotations Field object created using model's dataloader containing
+   *     annotation for images.
+   * @param output Output matrix where output will be stored.
```
Comment (KimSangYeon-DGU) [inline dataloader/preprocessor.hpp:104]: Hey @kartikdutt18, can you explain the input, output, and terms of `cell` and `grid` which are used in this implementation?

**Review by KimSangYeon-DGU — COMMENTED** [review]
Sorry for the late response. I left a question for me to get understood this feature.

File: dataloader/preprocessor.hpp (hunk 7 above)
Comment (kartikdutt18) [reply dataloader/preprocessor.hpp:104]: Hey @KimSangYeon-DGU, no worries, The input here would be a field type containing bounding boxes, the output would be an arma::mat which be would be of size batchSize * (featureMapSize) where featureSize would correspond to  S * S * (numBoundingBoxes * (numClasses + 5)) for YOLOv3 and S * S * (numBoundingBoxes * 5 + numClasses). Here cell is a single cell / data point in Feature Size. And the grid is featureSize reshaped as a cube (S, S, (numBoundingBoxes * 5 + numClasses)).

**Review by KimSangYeon-DGU — APPROVED** [review]
Looks good to me.

**Review by saksham189 — APPROVED** [review]

**Comment by saksham189** [comment]
Thanks for the great contribution @kartikdutt18 !

**Comment by kartikdutt18** [comment]
Thanks  lot @KimSangYeon-DGU, @saksham189, @zoq for the reviews and all the help. Also, thanks a lot for mentoring me during GSoC.
